flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -406,6 +409,7 @@ public WriterCallback(
         @Override
         public void onCompletion(RecordMetadata metadata, Exception exception) {
             if (exception != null) {
+                numRecordsOutErrorsCounter.inc();
```

Comment:
This seems to be some sort of new feature. Can you add it on a separate commit and a test for it?

Reply:
This is on the unhappy path that is normally not covered by ITCase. Do you have any idea on how to test it?

Reply:
One scenario that might work is that you create a second `FlinkKafkaInternalProducer` to fence the writer.

So you would first write some records to kafka with the writer and before calling `prepareCommit` create a FlinkKafkaInternalProducer that uses the same transaction id to writes some records and commits. I suspect that now calling `write` and `prepareCommit` on the writer lead to a `ProducerFencedException`.

Reply:
excellent!